server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -561,6 +561,33 @@ void innerUpdatePipelines(IngestMetadata newIngestMetadata) {
         }
     }
 
+    /**
+     * Determine if a pipeline contains a processor class within it by introspecting all of the processors within the pipeline.
+     * @param pipelineId the pipeline to inspect
+     * @param clazz the Processor class to look for
+     * @return True if the pipeline contains an instance of the Processor class passed in
+     */
+    public boolean hasProcessor(String pipelineId, Class<? extends Processor> clazz) {
+        Pipeline pipeline = getPipeline(pipelineId);
+        if (pipeline == null) {
+            return false;
+        }
+
+        for (Processor processor: pipeline.flattenAllProcessors()) {
+            if (processor instanceof WrappedProcessor) {
```

Comment:
I think this method does not work as expected in the following cases:
1) `hasProcessor("_id", ForEachProcessor.class/ConditionalProcessor.class)` invocations, because both classes implemented `WrappedProcessor`, we never end up checking whether `processor` is assignable from `clazz` and straight check the inner processor.
2) In the case that a wrapped processor is wrapped by another wrapped processor (for example: foreach processor wraps a `conditional` processor) then we never check the 2nd wrapped processor or any processor below that.

I think something like the following fixes the above cases:

```
if (clazz.isAssignableFrom(processor.getClass())) {
   return true;
}

while (processor instanceof WrappedProcessor) {
   WrappedProcessor wrappedProcessor = (WrappedProcessor) processor;
   if (clazz.isAssignableFrom(wrappedProcessor.getInnerProcessor().getClass())) {
      return true;
   }
   processor = wrappedProcessor.getInnerProcessor();
   if (wrappedProcessor == processor) {
      break;
   }
}
```

(replaces the body of the for loop)